README.md
Open Industrial ROS Controllers [![Build Status](https://travis-ci.org/start-jsk/open_industrial_ros_controllers.svg?branch=indigo-devel)](https://travis-ci.org/start-jsk/open_industrial_ros_controllers)
===============================



open_controllers_interface/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package open_controllers_interface
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.1.0 (2015-10-31)
------------------
* indent warning output
* minor fix
* add a boilerplait for recovering controller
* g_halt_motors\_ is not used
* do not use raw pointer
* follow standard c++ style class member naming
* add initializeCM virtual function for pre/post-processing by successor
* Contributors: Shohei Fujii

1.0.0 (2015-10-30)
------------------
* set default target to indigo
* remove old rosbuild settings
* Contributors: Kei Okada

0.1.4 (2013-12-13)
------------------
* remove unused %s, this causes segfault as reported on #14
* Contributors: Wataru Yasuda

0.1.3 (2013-12-07)
------------------

0.1.2 (2013-11-19)
------------------
* open_controllers_interface) install header to appropriate location (fix to https://github.com/start-jsk/open_industrial_ros_controllers/issues/11)
* Contributors: Isao Isaac Saito, Kei Okada

0.1.1 (2013-11-18)
------------------
* open_controllers_interface) Install headers.
* Contributors: Isao Isaac Saito

* Initial release


open_industrial_web_interface/www/js/basic.js

/**
 * Setup all visualization elements when the page is loaded. 
 */

var fixed_frame;
var current_group;
var link_group;
var end_effector_link;
var initial_flag = true;
var tfClient;

function init() {
    // Connect to ROS.
    var url = 'ws://' + location.hostname + ':9090';

    var real_ros = new ROSLIB.Ros({
        url : url
    });
    var virtual_ros = new ROSLIB.Ros({
        url : url
    });

    var joint_ros = new ROSLIB.Ros({
        url : url
    });

    joint_names = new ROSLIB.Param({
        ros: joint_ros,
        name: '/joint'
    });

    pub = new ROSLIB.Topic({
        ros: joint_ros,
        name: '/virtual_update_joint_position',
        messageType: 'std_msgs/Float64MultiArray'
    });

    moveit_pub = new ROSLIB.Topic({
        ros: joint_ros,
        name: '/moveit_joint',
        messageType: 'open_industrial_web_interface/MoveGroupPlan'
    });

    execute_pub = new ROSLIB.Topic({
        ros: joint_ros,
        name: '/execute_trajectory',
        messageType: 'std_msgs/Empty'
    });


    joint_pub = new ROSLIB.Topic({
        ros: joint_ros,
        name: '/update_joint_position',
        messageType: 'std_msgs/Float64MultiArray'
    });

    computefkClient = new ROSLIB.Service({
        ros : joint_ros,
        name : '/compute_fk',
        serviceType : 'moveit_msgs/GetPositionFK'
    });

    initial_interactive_pub = new ROSLIB.Topic({
        ros: joint_ros,
        name: '/initial_marker',
        messageType: 'std_msgs/String'
    });

    interactive_pub = new ROSLIB.Topic({
        ros: joint_ros,
        name: '/basic_marker/feedback',
        messageType: 'visualization_msgs/InteractiveMarkerFeedback'
    });

    plan_listener = new ROSLIB.Topic({
        ros: joint_ros,
        name: '/stock_joint_position',
        messageType: 'std_msgs/Float64MultiArray'
    });


    // Create the main viewer.
    var width = parseInt($("#main-content").css("width"));
    var height = document.height;

    var viewer = new ROS3D.Viewer({
        divID : 'urdf',
        width : width * 0.8,
        height : height * 0.8,
        antialias : true
    });

    // Add a grid.
    viewer.addObject(new ROS3D.Grid());


    fixed_frame_param = new ROSLIB.Param({
        ros: real_ros,
        name: '/fixed_frame'
    });

    link_group_param = new ROSLIB.Param({
        ros: virtual_ros,
        name: '/link_group/'
    });

    end_effector_link_param = new ROSLIB.Param({
        ros: joint_ros,
        name: '/end_effector_link/'
    });

    // Setup listener
    var listener = new ROSLIB.Topic({
        ros : virtual_ros,
        name : '/virtual_joint_states',
        messageType : 'sensor_msgs/JointState'
    });


    // Setup a client to listen to TFs.
    fixed_frame_param.get(function(value) {
        fixed_frame = value;
        tfClient = new ROSLIB.TFClient({
            ros : real_ros,
            fixedFrame : fixed_frame,
            angularThres : 0.01,
            transThres : 0.01,
            rate : 10.0
        });
        // Setup the marker client.
        var imClient = new ROS3D.InteractiveMarkerClient({
            ros : real_ros,
            tfClient : tfClient,
            topic : '/basic_marker',
            camera : viewer.camera,
            rootObject : viewer.selectableObjects
        });
    });

    link_group_param.get(function(value) {
        link_group = value;
        for (group_name in link_group) {
            $('#group').append("<option value=" + group_name + ">" + group_name + "</option>");
        }
        $("select#group").bind('change', function() {
            var selector = $("select#group option");
            selector.each(function() {
                $("#" + $(this).val()).hide();
            });
            var group = $("select#group option:selected").val();
            current_group = group;
            $("#" + group).show();
            var msg = new ROSLIB.Message({
                data: current_group
            });
            initial_interactive_pub.publish(msg);
            create_joint_position_msg(1, true);
        });
        setTimeout(function() {
            createSliderView();

            listener.subscribe(function(message) {
                for (i = 0; i < message.name.length; i++) {
                    var joint_name, joint_num;
                
                    for (j = 0; j < link_group[current_group].length; j++) {
                        if (link_group[current_group][j] == message.name[i]) {
                            var min = $('input#' + link_group[current_group][j]).attr("min");
                            var max = $('input#' + link_group[current_group][j]).attr("max");
                            var percent = parseInt((message.position[i] - min)/(max - min) * 100);
                            $('input#' + link_group[current_group][j]).attr("value", message.position[i]);
                            $('input#' + link_group[current_group][j]).next().children("div").attr("style", "width: " + percent + "%;");
                            $('input#' + link_group[current_group][j]).next().children("a").attr("aria-valuenow", message.position[i]);
                            $('input#' + link_group[current_group][j]).next().children("a").attr("aria-valuetext", 0.5);
                            $('input#' + link_group[current_group][j]).next().children("a").attr("title", message.position[i]);
                            $('input#' + link_group[current_group][j]).next().children("a").attr("style", "left: " +  percent + "%;");
                            $('input#' + link_group[current_group][j]).attr("value", message.position[i]);
                            break;
                        }
                    }
                }
            });

            create_joint_position_msg(1, true);
        }, 3000);

        setTimeout(function() {
            // Setup the URDF client.
            var virtualUrdfClient = new ROS3D.UrdfClient({
                ros : virtual_ros,
                tfPrefix : 'virtual',
                color : 0xff3000,
                tfClient : tfClient,
                param : 'robot_description',
                rootObject : viewer.scene
            });


            var urdfClient = new ROS3D.UrdfClient({
                ros : real_ros,
                tfClient : tfClient,
                param : 'robot_description',
                rootObject : viewer.scene
            });
        }, 1500);
    });

    end_effector_link_param.get(function(value) {
        end_effector_link = value;
    });

    plan_listener.subscribe(function(message) {
        message_stock.push(message);
    });

    $("button#moveit").click(function() {
        var msg = create_joint_position_msg(0,false);
        moveit_pub.publish(msg);
    });

    $("button#plan").click(function() {
        message_stock = new Array();
        var msg = create_joint_position_msg(0,true);
        moveit_pub.publish(msg);
    });

    $("button#execute").click(function() {
        if(message_stock != null) {
            sim_mode = new ROSLIB.Param({
                ros: joint_ros,
                name: '/sim_mode'
            });
            sim_mode.get(function(value) {
                if (value == true) {
                    timer = setInterval("joint_publish()",100);
                }
                else {
                    var msg = new ROSLIB.Message({
                    });
                    execute_pub.publish(msg);
                }                
            });
        }
    });
}

function create_joint_position_msg(type, plan_only) {

    positions = new Array();
    joints = new Array();
    dims = new Array();

    $("#" + current_group).children("label").each(function() {
        var dim = new ROSLIB.Message({
            label: ($(this).attr("id").split("-")[0]),
            size: ($(this).attr("id").split("-")[0]).length,
            stride: ($(this).attr("id").split("-")[0]).length
        });
        dims.push(dim);
        joints.push($(this).attr("id").split("-")[0]);
        positions.push(parseFloat($(this).next().children("input").next().children("a").attr("aria-valuenow")));
    });

    var msg;
    if (type == 0) {
        msg = new ROSLIB.Message({
            joint: {
                layout: {
                    dim: dims,
                    data_offset: 0
                },
                data: positions
            },
            plan_only: plan_only,
            group_name: current_group
        });
    }
    else {
        msg = new ROSLIB.Message({
            layout: {
                dim: dims,
                data_offset: 0
            },
            data: positions
        });
    }
    var fk_link_name;

    if (end_effector_link[current_group] == null) {
        estimated_link_name = joints[joints.length - 1];
        fk_link_name = estimated_link_name.charAt(0).toUpperCase() + estimated_link_name.slice(1);
    }
    else {
        fk_link_name = end_effector_link[current_group];
    }

    // Update interactive marker poisition
    var request = new ROSLIB.ServiceRequest({
        header: {
            seq: 0,
            stamp: 0,
            frame_id: fixed_frame
        },
        fk_link_names: [fk_link_name],
        robot_state: {
            joint_state: {
                name: joints,
                position: positions
            }
        }
    });

    computefkClient.callService(request, function(result) {

        var interactive_msg = new ROSLIB.Message({
            marker_name: "basic",
            event_type: 5,
            pose: {
                position: {
                    x: result.pose_stamped[0].pose.position.x,
                    y: result.pose_stamped[0].pose.position.y,
                    z: result.pose_stamped[0].pose.position.z
                },
                orientation: {
                    x: result.pose_stamped[0].pose.orientation.x,
                    y: result.pose_stamped[0].pose.orientation.y,
                    z: result.pose_stamped[0].pose.orientation.z,
                    w: result.pose_stamped[0].pose.orientation.w
                }
            }
        });
        interactive_pub.publish(interactive_msg);
    });
    return msg;
}

function callback() {
    var msg = create_joint_position_msg(1, true);
    pub.publish(msg);
}

function joint_publish() {
    if(message_stock.length == 0) {
        clearInterval(timer);
    }
    else {
        joint_pub.publish(message_stock.shift());
    }
}

// create joint_publisher
function createSliderView() {
    var i = 0;
    for (group_name in link_group) {
        $("#slider-pane").append('<div id="' + group_name + '"/>');
        if (i != 0) {
            $("#" + group_name).hide();
        }
        else {
            current_group = group_name;
            i++;
        }
    }
    joint_names.get(function(value) {
        names = value.names;
        for (group_name in link_group) {
            for (var i = 0;i < names.length;i++) {
                if (link_group[group_name].indexOf(names[i]) != -1) {
                    child = $('<label>', {for: names[i], text: names[i]});
                    child2 = $('<input>', {type: "range", name: names[i], id: names[i], value: 0, max: eval("value." + names[i] + ".max"), min: eval("value." + names[i] + ".min"), step: 0.000001, "data-highlight": "true", "data-mini": "true",
                                   onchange: "callback()"});
                    $("#" + group_name).append(child);
                    $("#" + group_name).append(child2);
                }
            }
        }
        $.getScript("js/jquery-mobile/jquery.mobile-1.3.2.min.js");
        var msg = new ROSLIB.Message({
            data: current_group
        });
        initial_interactive_pub.publish(msg);

    });
}


open_industrial_web_interface/nodes/moveit_publisher.py
#!/usr/bin/python

import roslib
import rospy
import time
import actionlib
roslib.load_manifest("open_industrial_web_interface")

from open_industrial_web_interface.msg import MoveGroupPlan
from moveit_msgs.srv import *
from moveit_msgs.msg import RobotState
from moveit_msgs.msg import WorkspaceParameters
from moveit_msgs.msg import JointConstraint
from moveit_msgs.msg import Constraints
from moveit_msgs.msg import MoveGroupAction
from moveit_msgs.msg import MoveGroupGoal
from moveit_msgs.msg import MoveGroupActionGoal
from moveit_msgs.msg import MotionPlanRequest
from geometry_msgs.msg import Vector3
from std_msgs.msg import Float64MultiArray
from std_msgs.msg import Empty
from std_msgs.msg import MultiArrayDimension
from sensor_msgs.msg import JointState
from moveit_msgs.msg import MoveGroupActionResult

def joint_position_callback(joints):

    global joint_states, plan_only
    fixed_frame = rospy.get_param("/fixed_frame")

    client = actionlib.SimpleActionClient('move_group', MoveGroupAction)
    client.wait_for_server()
    move_group_goal = MoveGroupGoal()

    try:

        msg = MotionPlanRequest()
        workspace_parameters = WorkspaceParameters()
        workspace_parameters.header.stamp = rospy.Time.now()
        workspace_parameters.header.frame_id = fixed_frame
        workspace_parameters.min_corner = Vector3(-1.0, -1.0, -1.0)
        workspace_parameters.max_corner = Vector3(1.0, 1.0, 1.0)

        start_state = RobotState()
#        start_state.joint_state.header.stamp = rospy.Time.now()
        start_state.joint_state.header.frame_id = fixed_frame
        start_state.joint_state.name =  []
        start_state.joint_state.position = []

        cons = Constraints()
        cons.name = ""
        i = 0
        for dim in joints.joint.layout.dim:
            start_state.joint_state.name.append(dim.label)
            if dim.label == "r_elbow_flex_joint": #tmp
                start_state.joint_state.position.append(-0.1)
            elif dim.label == "l_elbow_flex_joint": #tmp
                start_state.joint_state.position.append(-0.1)
            else:
                start_state.joint_state.position.append(joint_states[dim.label])
            jc = JointConstraint()
            jc.joint_name = dim.label
            jc.position = joints.joint.data[i]
            jc.tolerance_above = 0.0001
            jc.tolerance_below = 0.0001
            jc.weight = 1.0
            i = i + 1
            cons.joint_constraints.append(jc)


        msg.workspace_parameters = workspace_parameters
        msg.start_state = start_state
        msg.goal_constraints.append(cons)

        msg.num_planning_attempts = 1
        msg.allowed_planning_time = 5.0

        msg.group_name = joints.group_name
        move_group_goal.request = msg


        if joints.plan_only:
            plan_only = True
            move_group_goal.planning_options.plan_only = True
        else:
            plan_only = False

        client.send_goal(move_group_goal)
        client.wait_for_result(rospy.Duration.from_sec(5.0))


    except rospy.ROSInterruptException, e:
        print "failed: %s"%e


def moveit_callback(msg):
    pub = rospy.Publisher('/update_joint_position', Float64MultiArray)
    stock_pub = rospy.Publisher('/stock_joint_position', Float64MultiArray)

    global plan_only

    r = rospy.Rate(10)
    names = msg.result.planned_trajectory.joint_trajectory.joint_names
    points = msg.result.planned_trajectory.joint_trajectory.points

    for point in points:
        pos_msg = Float64MultiArray()
        i = 0
        for name in names:
            dim = MultiArrayDimension()
            dim.label = name
            pos_msg.layout.dim.append(dim)
        pos_msg.data = point.positions
        if rospy.get_param('~sim_mode') and plan_only == False:
            pub.publish(pos_msg)
        elif plan_only:
            stock_pub.publish(pos_msg)
        r.sleep()

    if plan_only:
        global robot_trajectory
        robot_trajectory = msg.result.planned_trajectory

def execute_callback(msg):
    global robot_trajectory
    rospy.wait_for_service('execute_kinematic_path')
    try:
        execute_service = rospy.ServiceProxy('execute_kinematic_path', ExecuteKnownTrajectory)
        srv = ExecuteKnownTrajectoryRequest()
        srv.trajectory = robot_trajectory
        response = execute_service(srv)
        print response.error_code
    except rospy.ServiceException, e:
        print "Service call failed: %s"%e

def joint_state_callback(msg):
    global joint_states
    i = 0
    for name in msg.name:
        joint_states[name] = msg.position[i]
        i = i + 1

if __name__ == '__main__':
    global joint_states
    joint_states = {}
    rospy.init_node('moveit_publisher')
    rospy.Subscriber("/execute_trajectory", Empty, execute_callback)
    rospy.Subscriber("moveit_joint", MoveGroupPlan, joint_position_callback)
    rospy.Subscriber("/move_group/result", MoveGroupActionResult, moveit_callback)
    rospy.Subscriber("/joint_states", JointState, joint_state_callback)
    rospy.spin()


open_controllers_interface/include/open_controllers_interface/open_controllers_interface.h
// -*- mode: c++ -*-
#ifndef __OPEN_CONTROLLERS_INTERFACE_H__
#define __OPEN_CONTROLLERS_INTERFACE_H__
#include <iostream>
#include "ros/ros.h"
#include <std_srvs/Empty.h>
#include <boost/shared_ptr.hpp>
#include <boost/accumulators/accumulators.hpp>
#include <boost/accumulators/statistics/stats.hpp>
#include <boost/accumulators/statistics/max.hpp>
#include <boost/accumulators/statistics/mean.hpp>
#include <boost/thread.hpp>
#include <diagnostic_updater/DiagnosticStatusWrapper.h>
#include <pr2_controller_manager/controller_manager.h>
#include <diagnostic_msgs/DiagnosticArray.h>

#include <ros/ros.h>
#include <std_srvs/Empty.h>

#include <realtime_tools/realtime_publisher.h>
#include <std_msgs/Float64.h>


namespace OpenControllersInterface {
  // if groovy
  //typedef std::vector<pr2_mechanism_model::Transmission>::iterator TransmissionIterator;
  // if hydro
  typedef std::vector<boost::shared_ptr<pr2_mechanism_model::Transmission> >::iterator TransmissionIterator;
  //
  typedef boost::accumulators::accumulator_set<double, boost::accumulators::stats<boost::accumulators::tag::max, boost::accumulators::tag::mean> > DoubleAccumulatorSet;
  struct Stat {
    DoubleAccumulatorSet ec_acc;
    DoubleAccumulatorSet cm_acc;
    DoubleAccumulatorSet loop_acc;
    DoubleAccumulatorSet jitter_acc;
    int overruns;
    int recent_overruns;
    int last_overrun;
    int last_severe_overrun;
    unsigned long loop_count;
    double overrun_loop_sec;
    double overrun_ec;
    double overrun_cm;
    
    // These values are set when realtime loop does not meet performace expections
    bool rt_loop_not_making_timing; 
    double halt_rt_loop_frequency;
    double rt_loop_frequency;
  };


  class ControllerStatus {
    public:
    ControllerStatus(){}
    virtual ~ControllerStatus(){}
    virtual bool isHealthy(){return true;}
  };
  typedef boost::shared_ptr<ControllerStatus> ControllerStatusPtr;

  class OpenController {
  public:
    OpenController();
    virtual ~OpenController();
    virtual ControllerStatusPtr updateJoints(struct timespec*) = 0;
    virtual ControllerStatusPtr recoverController() = 0;
    virtual void finalizeHW() = 0;

    static bool initRT();
    
    void setAllowUnprogrammedP(bool val) {
      allow_unprogrammed_p_ = val;
    }
    
    void setStatsPublishP(bool val) {
      stats_publish_p_ = val;
    }
    
    void setRobotXMLFile(std::string val) {
      robot_xml_file_ = val;
    }

    bool setDryRun(bool _dryrun) {
      dryrunp_ = _dryrun;
      return _dryrun;
    }

    void start();

    int waitThreadJoin() {
      return 0;
    }

    // calling haltMotorsService
    virtual void quitRequest();

    bool resetMotorsService(std_srvs::Empty::Request &req,
          std_srvs::Empty::Response &resp);
    bool haltMotorsService(std_srvs::Empty::Request &req,
         std_srvs::Empty::Response &resp);
    bool publishTraceService(std_srvs::Empty::Request &req,
           std_srvs::Empty::Response &resp);
    
    // virtual function you need to impelement
    void initialize();
    void loadRobotDescription();
    void startMain();
    void finalize();

    void diagnosticLoop();

    int setupPidFile ();
    void cleanupPidFile ();
    void parseArguments(int argc, char** argv);
    void Usage(std::string msg = "");
  protected:
    int lock_fd(int fd);
    virtual void initializeROS(ros::NodeHandle& node) = 0;
    virtual void initializeCM() = 0;
    virtual void initializeHW() = 0;
    
    void publishDiagnostics();
    double now();
    void timespecInc(struct timespec &tick, int nsec);
    double publishJitter(double start);

    std::string piddir_;
    std::string pidfile_;
    bool dryrunp_;
    bool not_sleep_clock_;
    bool allow_unprogrammed_p_;
    bool stats_publish_p_;
    bool g_reset_motors_;
    bool g_quit_;;
    bool g_halt_requested_;
    bool g_publish_trace_requested_;
    std::string robot_xml_file_;
    double min_acceptable_rt_loop_frequency_;
    double period_;
    std::string g_robot_desc_;
    pr2_hardware_interface::HardwareInterface* hw_;
    boost::shared_ptr<pr2_controller_manager::ControllerManager> cm_;
    boost::shared_ptr<realtime_tools::RealtimePublisher<std_msgs::Float64> > rtpublisher_;
    boost::shared_ptr<realtime_tools::RealtimePublisher<diagnostic_msgs::DiagnosticArray> > publisher_;
    Stat g_stats_;

    ros::ServiceServer reset_service_;
    ros::ServiceServer halt_service_;
    ros::ServiceServer publishTrace_service_;
    
  private:
  };

  class RTLoopHistory {
  public:
    RTLoopHistory(unsigned length, double default_value) :
      index_(0), 
      length_(length),
      history_(new double[length]) {
      for (unsigned i=0; i<length_; ++i) 
        history_[i] = default_value;
    }

    ~RTLoopHistory() {
      delete[] history_;
      history_ = NULL;
    }
  
    void sample(double value) {
      index_ = (index_+1) % length_;
      history_[index_] = value;
    }

    double average() const {
      double sum(0.0);
      for (unsigned i=0; i<length_; ++i) 
        sum+=history_[i];
      return sum / double(length_);
    }

  protected:
    unsigned index_;
    unsigned length_;
    double *history_;
  };

  class Finalizer {
  protected:
    OpenController* controller;
  public:
    Finalizer(OpenController* controller_): controller(controller_) { };
    virtual ~Finalizer();
  };

}
#endif


open_industrial_web_interface/nodes/interactive_moveit.py
#!/usr/bin/python

import roslib
import rospy
from geometry_msgs.msg import PoseStamped
from moveit_msgs.srv import *
from moveit_msgs.msg import RobotState
from std_msgs.msg import Float64MultiArray
from std_msgs.msg import String
from std_msgs.msg import MultiArrayDimension
from sensor_msgs.msg import JointState
from interactive_markers.interactive_marker_server import (
    Marker,
    InteractiveMarker,
    InteractiveMarkerServer,
    InteractiveMarkerControl,
    InteractiveMarkerFeedback)
server = None

def makeInteractiveMarker(name, description):
    global fixed_frame
    interactive_marker = InteractiveMarker()
    interactive_marker.header.frame_id = fixed_frame
    interactive_marker.name = name
    interactive_marker.description = description
    return interactive_marker


def makeInteractiveMarkerControl(interactive_marker, mode):
    interactive_marker_control = InteractiveMarkerControl()
    interactive_marker_control.always_visible = True
    interactive_marker_control.interaction_mode = mode
    interactive_marker.controls.append(interactive_marker_control)
    return interactive_marker_control

def setOrientation(w, x, y, z, marker):
    marker.orientation.w = w
    marker.orientation.x = x
    marker.orientation.y = y
    marker.orientation.z = z

def setColor(red, green, blue, alpha, marker):
    marker.color.r = red
    marker.color.g = green
    marker.color.b = blue
    marker.color.a = alpha


def callback(req):
    print req
    return GetPositionIKResponse

def initial_callback(msg):
    print "init"
    global joint_states,joint_names, initial_joint_position, group_name
    group_name = msg.data
    joint_names = rospy.get_param("/link_group/" + group_name)
    initial_joint_position = [0] * len(joint_names)


def joint_state_callback(msg):
    global joint_states
    joint_states = msg

def feedback(feedback):
    global pub, initial_joint_position, fixed_frame, joint_names, group_name

    server.setPose(feedback.marker_name, feedback.pose)
    server.applyChanges()

    rospy.wait_for_service('compute_ik')
    try:
        service = rospy.ServiceProxy('compute_ik', GetPositionIK)
        request = GetPositionIKRequest()
        request.ik_request.group_name = group_name
        request.ik_request.timeout = rospy.Duration.from_sec(0.0001)

        # initial robot state
        robot_state = RobotState()
        robot_state.joint_state.header.frame_id = fixed_frame
        robot_state.joint_state.name =  joint_names
        robot_state.joint_state.position = initial_joint_position
        robot_state.joint_state.velocity =  []
        request.ik_request.robot_state = robot_state

        # goal end pose
        pose_stamped = PoseStamped()
        pose_stamped.header.frame_id = fixed_frame
        pose_stamped.pose.position.x = feedback.pose.position.x
        pose_stamped.pose.position.y = feedback.pose.position.y
        pose_stamped.pose.position.z = feedback.pose.position.z
        pose_stamped.pose.orientation.x = feedback.pose.orientation.x
        pose_stamped.pose.orientation.y = feedback.pose.orientation.y
        pose_stamped.pose.orientation.z = feedback.pose.orientation.z
        pose_stamped.pose.orientation.w = feedback.pose.orientation.w

        request.ik_request.pose_stamped = pose_stamped
        response = service(request)
        print response
        if len(response.solution.joint_state.position) != 0:
            print "success"
            msg = Float64MultiArray()
            for i,joint_name in enumerate(response.solution.joint_state.name):
                for j, name in enumerate(joint_names):
                    if joint_name == name:
                        initial_joint_position[j] = response.solution.joint_state.position[i]
                        dim = MultiArrayDimension()
                        dim.label = name
                        msg.layout.dim.append(dim)
                        msg.data.append(response.solution.joint_state.position[i])
            pub.publish(msg)

    except rospy.ServiceException, e:
        print "Service call failed: %s"%e




if __name__ == '__main__':
    rospy.init_node("moveit_interactive_marker")
    global pub, fixed_frame
    fixed_frame = rospy.get_param("/fixed_frame")

    pub = rospy.Publisher('virtual_update_joint_position',Float64MultiArray)
    rospy.Subscriber('/initial_marker', String, initial_callback)
    rospy.Subscriber('/virtual_joint_states', JointState, joint_state_callback)

    server = InteractiveMarkerServer("basic_marker")

    interactive_marker = makeInteractiveMarker("basic", "")
    interactive_marker.scale = 0.3
    interactive_marker.pose.position.x = 0
    interactive_marker.pose.position.y = 0
    interactive_marker.pose.position.z = 1.0

    control_slide_x = makeInteractiveMarkerControl(
        interactive_marker, InteractiveMarkerControl.MOVE_AXIS)
    control_slide_y = makeInteractiveMarkerControl(
        interactive_marker, InteractiveMarkerControl.MOVE_AXIS)
    control_slide_z = makeInteractiveMarkerControl(
        interactive_marker, InteractiveMarkerControl.MOVE_AXIS)
    control_rotate_x = makeInteractiveMarkerControl(
        interactive_marker, InteractiveMarkerControl.ROTATE_AXIS)
    control_rotate_y = makeInteractiveMarkerControl(
        interactive_marker, InteractiveMarkerControl.ROTATE_AXIS)
    control_rotate_z = makeInteractiveMarkerControl(
        interactive_marker, InteractiveMarkerControl.ROTATE_AXIS)
    control_sphere = makeInteractiveMarkerControl(
        interactive_marker, InteractiveMarkerControl.MOVE_ROTATE_3D)
    marker = Marker()
    marker.color.r = 0.2
    marker.color.g = 0.3
    marker.color.b = 0.7
    marker.color.a = 0.5

    marker.type = Marker.SPHERE
    marker.scale.x = 0.2
    marker.scale.y = 0.2
    marker.scale.z = 0.2
    control_sphere.markers.append(marker)

    setOrientation(1,1,0,0, control_slide_x)
    setOrientation(1,0,1,0, control_slide_y)
    setOrientation(1,0,0,1, control_slide_z)
    setOrientation(1,1,0,0, control_rotate_x)
    setOrientation(1,0,1,0, control_rotate_y)
    setOrientation(1,0,0,1, control_rotate_z)
    setOrientation(1,1,0,0, control_sphere)


    server.insert(interactive_marker, feedback)
    server.applyChanges()
    rospy.spin()


